Upgrade to spack-stack on non-production machines

DavidHuber-NOAA opened the pull request:
This upgrades the modules for gfs-utils to spack-stack/1.5.1.  This has been tested thoroughly on Hera and somewhat on Orion.  Builds were tested on S4, Jet, and WCOSS2.  Backwards compatibility had to be added into the `CMakeLists.txt` and source files for `mkgfsawps` and `fbwndgfs` applications for `w3emc` and `ip` to continue support for WCOSS2.

Resolves #23

aerorahul commented on src/enkf_chgres_recenter_nc.fd/CMakeLists.txt:
```diff
@@ -8,7 +8,10 @@ utils.f90
-  set(CMAKE_Fortran_FLAGS "${CMAKE_Fortran_FLAGS} -fp-model precise")
+  set(CMAKE_Fortran_FLAGS "${CMAKE_Fortran_FLAGS} -fp-model precise -fpp")
+endif()
+if(ip_VERSION GREATER_EQUAL 4.0.0)
+  set(CMAKE_Fortran_FLAGS "${CMAKE_Fortran_FLAGS} -DIP_V4")
```
I think `target_compile_definitions(${exe_name} PRIVATE  "IP_V4")` should be used before the `target_link_libraries` line below.

You shouldn't then need the Fortran preprocessor (`-fpp`) compiler option.

DavidHuber-NOAA replied:
@aerorahul When I try this I receive the cmake error
```
CMake Error at src/enkf_chgres_recenter_nc.fd/CMakeLists.txt:19 (target_compile_definitions):
  Cannot specify compile definitions for target "exe_name" which is not built
  by this project.
```

DavidHuber-NOAA replied:
Oh, I see it.  It should be ${exe_name}.

aerorahul replied:
yeah. I updated my lazy comment.

DavidHuber-NOAA replied:
But using this seems to dis-envoke the preprocessor without `-fpp`:

```
/scratch1/NCEPDEV/nems/David.Huber/GW/gw_spack-stack_151/sorc/gfs_utils.fd/src/enkf_chgres_recenter_nc.fd/interp.f90(210): warning #5117: Bad # preprocessor line
#ifdef IP_V4
-^
/scratch1/NCEPDEV/nems/David.Huber/GW/gw_spack-stack_151/sorc/gfs_utils.fd/src/enkf_chgres_recenter_nc.fd/interp.f90(212): warning #5117: Bad # preprocessor line
#endif
```

aerorahul replied:
Hmm.  I have never had that problem and the documentation doesn't seem to mention this:
https://cmake.org/cmake/help/latest/command/target_compile_definitions.html#command:target_compile_definitions

aerorahul replied:
I know.  Rename the source code to `.F90`.

aerorahul replied:
CMake will invoke the preprocessor automatically if the extension is `.F90` to create the `.f90` file prepared for the compiler.

aerorahul commented on src/fbwndgfs.fd/CMakeLists.txt:
```diff
@@ -3,10 +3,17 @@ list(APPEND fortran_src
+string(REPLACE "." ";" W3EMC_VERSIONS ${w3emc_VERSION})
+list(GET W3EMC_VERSIONS 0 W3EMC_MAJOR)
+list(GET W3EMC_VERSIONS 1 W3EMC_MINOR)
+if(W3EMC_MINOR GREATER_EQUAL 10 AND W3EMC_MAJOR GREATER_EQUAL 2)
+  set(CMAKE_Fortran_FLAGS "${CMAKE_Fortran_FLAGS} -DW3EMC_210")
+endif()
```
same comment as in the `enkf_chgres_recenter_nc.fd/CMakeLists.txt` addressing the compiler preprocessor

aerorahul commented on src/mkgfsawps.fd/CMakeLists.txt:
```diff
@@ -3,19 +3,22 @@ list(APPEND fortran_src
   set_source_files_properties(${fortran_src_free} PROPERTIES COMPILE_FLAGS "-ffree-form")
 endif()
+if(ip_VERSION GREATER_EQUAL 4.0.0)
+  set(CMAKE_Fortran_FLAGS "${CMAKE_Fortran_FLAGS} -DIP_V4")
```
and here.

aerorahul reviewed (changes requested):
Suggestions to properly invoke preprocessor and provide compile time definitions.

aerorahul commented on src/fbwndgfs.fd/CMakeLists.txt:
```diff
@@ -3,10 +3,17 @@ list(APPEND fortran_src
 elseif(CMAKE_Fortran_COMPILER_ID MATCHES "^(GNU)$")
   set(CMAKE_Fortran_FLAGS "${CMAKE_Fortran_FLAGS} -fdefault-real-8 -fdefault-integer-8 -fconvert=big-endian")
 endif()
+#Parse w3emc version numbers
+string(REPLACE "." ";" W3EMC_VERSIONS ${w3emc_VERSION})
+list(GET W3EMC_VERSIONS 0 W3EMC_MAJOR)
+list(GET W3EMC_VERSIONS 1 W3EMC_MINOR)
+if(W3EMC_MINOR GREATER_EQUAL 10 AND W3EMC_MAJOR GREATER_EQUAL 2)
```
```suggestion
if(w3emc_VERSION VERSION_GREATER 2.10)
```

aerorahul commented on src/enkf_chgres_recenter_nc.fd/CMakeLists.txt:
```diff
@@ -8,7 +8,10 @@ utils.f90
-  set(CMAKE_Fortran_FLAGS "${CMAKE_Fortran_FLAGS} -fp-model precise")
+  set(CMAKE_Fortran_FLAGS "${CMAKE_Fortran_FLAGS} -fp-model precise -fpp")
+endif()
+if(ip_VERSION GREATER_EQUAL 4.0.0)
```
```suggestion
if(ip_VERSION VERSION_GREATER_EQUAL 4.0.0)
```
I think we should use the CMake's version compare rather than string compare.
See the [if](https://cmake.org/cmake/help/v3.20/command/if.html) documentation

aerorahul commented on src/fbwndgfs.fd/CMakeLists.txt:
```diff
@@ -7,11 +7,7 @@ if(CMAKE_Fortran_COMPILER_ID MATCHES "^(Intel)$")
-list(GET W3EMC_VERSIONS 0 W3EMC_MAJOR)
-list(GET W3EMC_VERSIONS 1 W3EMC_MINOR)
-if(W3EMC_MINOR GREATER_EQUAL 10 AND W3EMC_MAJOR GREATER_EQUAL 2)
+if(w3emc_VERSION VERSION GREATER 2.10)
```
```suggestion
if(w3emc_VERSION VERSION_GREATER_EQUAL 2.10)
```
Apologies.  I made a mistake.

aerorahul commented on src/mkgfsawps.fd/CMakeLists.txt:
```diff
@@ -3,19 +3,22 @@ list(APPEND fortran_src
   set(CMAKE_Fortran_FLAGS "${CMAKE_Fortran_FLAGS} -fconvert=big-endian")
   set_source_files_properties(${fortran_src_free} PROPERTIES COMPILE_FLAGS "-ffree-form")
 endif()
+if(ip_VERSION GREATER_EQUAL 4.0.0)
```
```suggestion
if(ip_VERSION VERSION_GREATER_EQUAL 4.0.0)
```

DavidHuber-NOAA commented:
@aerorahul Thanks for the suggestions!  I reworked the CMakeLists.txt files to better check for version numbers and removed the `-fpp` flags while renaming the three source files to use `.F` or `.F90` extensions where preprocessing was needed.  This compiled on both Hera and WCOSS2.

aerorahul commented on src/enkf_chgres_recenter_nc.fd/CMakeLists.txt:
```diff
@@ -10,6 +10,9 @@ utils.f90
 if(CMAKE_Fortran_COMPILER_ID MATCHES "^(Intel)$")
   set(CMAKE_Fortran_FLAGS "${CMAKE_Fortran_FLAGS} -fp-model precise")
 endif()
+if(ip_VERSION VERSION_GREATER_EQUAL 4.0.0)
+  set(CMAKE_Fortran_FLAGS "${CMAKE_Fortran_FLAGS} -DIP_V4")
+endif()
```
Was `target_compile_definitions` not working? 
`CMAKE_Fortran_FLAGS` are typically compiler flags, while compile definitions are preprocessor directives.

DavidHuber-NOAA replied:
Whoops, I got a little type-happy and removed that.  I've switch back which compiles on both Hera and WCOSS2.

aerorahul reviewed (approved):
Looks good.
Thank you for including the suggested changes.
